Where's Waldo Blog Post (#133)
risc0/website: blog/2023-03-XX-waldo/index.md

### Image Manipulation

In order to manipulate the image and cut-out Waldo, and in particular to crop and apply a mask, this example utilizes the popular `image` crate.
This is enabled by implementing `image::GenericImageView` on `ImageOracle`.
With that trait, many of the image operations provided in the `image` crate, and by [others](https://docs.rs/imageproc/latest/imageproc/), can be used on `ImageOracle` inside the guest.
A similar approach could be used to produce a provable blur, image down-scaling, and more.

One the most powerful aspects of RISC Zero is the ability to use your favorite libraries, and avoid reinventing the wheel.

### Get Involved

Just as the current ecosystem of amazing open-source applications is powered on amazing open-source libraries, so will be the zkVM programs you write.
And if you ever try to import a library and it doesn't work for some reason, it would be greatly appreciated if you let us know by filing an issue on [GitHub](https://github.com/risc0/risc0).

## Run this example

> You can find the source code and build instructions for this example (and many others) in the [risc0/examples folder on GitHub](https://github.com/risc0/risc0/tree/main/examples). Run it yourself, and use it as a starting point to to create your own projects! 

Reviewer: ```suggestion
You can find the source code and build instructions for this example (and many others) in the [risc0/examples folder on GitHub](https://github.com/risc0/risc0/tree/main/examples). Run it yourself, and use it as a starting point to to create your own projects! 

Note: This example is memory-intensive; we recommend using a machine with at least 64GB of RAM.
```

Reviewer 2: Hmm, will run it again on my machine later today b/c I thought it was running fine locally, and I "only" have 32GB -- but maybe I'm misremembering.

Otherwise the comment looks good, I'll just come back and update the number if needed...

Reviewer 2: good as-is, see https://github.com/risc0/risc0/pull/464

In particular, if you want to prove you know where Waldo is, without revealing his location, you can:

1. Have your friend, the verifier, print a Where’s Waldo puzzle on a big poster board.
2. Take the poster board, turn around and cut out Waldo from the image. Dispose of the rest.
3. Hand the cut out of Waldo to your friend, who makes sure it’s really Waldo.

This is *kind of* like a zero-knowledge proof (with less cryptography and more craft supplies), but why not implement a real zero-knowledge proof to show you know where Waldo is?

This example implements a RISC Zero program which allows a prover to convince a verifier they know Waldo’s location in a public Where’s Waldo puzzle, without revealing Waldo’s coordinates.

To get a better understanding, browse the [source code](https://github.com/risc0/risc0/tree/main/examples/waldo) and run the example yourself. 

## Overview

This example of implementing a proof for Where’s Waldo is designed to help you understand some useful techniques in creating more advanced programs with RISC Zero.

In particular, it should help you understand:

> - Using I/O to have the host provide the guest data on request.

Reviewer 2: Should we drop this line now that the I/O section is cut?
```suggestion
```

These modules implement the ideas described above.

Similar Merkle tree abstractions can be used to, for example, ensure a secret word is part of a dictionary, a payment destination is not in a list of banned addresses, or that a user is in the set of authorized users.

### Image Manipulation

In order to manipulate the image and cut-out Waldo, and in particular to crop and apply a mask, this example utilizes the popular `image` crate.
This is enabled by implementing `image::GenericImageView` on `ImageOracle`.
With that trait, many of the image operations provided in the `image` crate, and by [others](https://docs.rs/imageproc/latest/imageproc/), can be used on `ImageOracle` inside the guest.
A similar approach could be used to produce a provable blur, image down-scaling, and more.

One the most powerful aspects of RISC Zero is the ability to use your favorite libraries, and avoid reinventing the wheel.

### Get Involved

Just as the current ecosystem of amazing open-source applications is powered on amazing open-source libraries, so will be the zkVM programs you write.

> And if you ever try to import a library and it doesn't work for some reason, it would be greatly appreciated it you get us know by filing an issue on [GitHub](https://github.com/risc0/risc0).

Reviewer 2: Typos
```suggestion
And if you ever try to import a library and it doesn't work for some reason, it would be greatly appreciated if you let us know by filing an issue on [GitHub](https://github.com/risc0/risc0).
```

# Understanding Waldo

[Where’s Waldo](https://en.wikipedia.org/wiki/Where%27s_Wally%3F) is a [favorite analogy](https://medium.com/swlh/a-zero-knowledge-proof-for-wheres-wally-930c21e55399) for zero-knowledge proofs.
In particular, if you want to prove you know where Waldo is, without revealing his location, you can:

1. Have your friend, the verifier, print a Where’s Waldo puzzle on a big poster board.
2. Take the poster board, turn around and cut out Waldo from the image. Dispose of the rest.
3. Hand the cut out of Waldo to your friend, who makes sure it’s really Waldo.

This is *kind of* like a zero-knowledge proof (with less cryptography and more craft supplies), but why not implement a real zero-knowledge proof to show you know where Waldo is?

This example implements a RISC Zero program which allows a prover to convince a verifier they know Waldo’s location in a public Where’s Waldo puzzle, without revealing Waldo’s coordinates.

> You can find the source code for this example in the [risc0-rust-examples GitHub repository](Understanding%20Waldo%20f60dd196187241c99aa9bdee885cea54.md).

Reviewer: I think we're moving to `risc0/examples` rather than `risc0-rust-examples`, but @tzerrell is more clued in to the status of that transition.

Reviewer 2: Agreed with @pdg744, IMO this link should be to `risc0/examples/waldo`
```suggestion
You can find the source code for this example in the [waldo project in our examples directory](https://github.com/risc0/risc0/tree/main/examples/waldo).
```

> You can find the source code for this example in the [risc0-rust-examples GitHub repository](Understanding%20Waldo%20f60dd196187241c99aa9bdee885cea54.md).

Reviewer: (also link seems dead on my end)

> Browse the code, and run the example yourself, as your read to get a solid understanding for its parts.

Reviewer: Typo in `as your read`, and the flow of this sentence feels a bit awkward to me. Perhaps: 

"To get a solid understanding of this article, browse the code and run the example yourself."

Although inviting people to run the example themselves may merit some warning re: memory usage and build times... See https://github.com/risc0/risc0-rust-examples/issues/53
